### PR TITLE
Query::get_multiple and get_multiple_mut

### LINKED ISSUE #2042: Safe method for multiple mutable references to components from Query
## What problem does this solve or what need does it fill?

It is sometimes useful/easier to have mutable access to the same component of many different entities at once.

## What solution would you like?

Two new methods for Query, a default and one for a component. Example:

```rust
fn get_many_mut<'a, Q: WorldQuery, F: WorldQuery>(
    entities: impl IntoIterator<Item = Entity>,
    query: &'a mut Query<Q, F>,
) -> Result<Vec<<Q::Fetch as Fetch<'a>>::Item>, MultiEntityQueryErr>
where
    F::Fetch: FilterFetch,
{
    let mut unique = HashSet::default();
    let mut iter = entities.into_iter();
    if iter.by_ref().all(|x| unique.insert(x)) {
        let mut result = Vec::with_capacity(unique.len());

        for entity in iter {
            match unsafe { query.get_unchecked(entity) } {
                Ok(item) => {
                    result.push(item);
                }
                Err(err) => {
                    return Err(MultiEntityQueryErr::QueryEntityError(err));
                }
            }
        }

        Ok(result)
    } else {
        Err(MultiEntityQueryErr::DuplicateEntities)
    }
}
```

Just a quick implementation I've already written, not necessarily the best. Originally I used const generic arrays to avoid heap allocation too - but depends on use case.

## What alternative(s) have you considered?

Have users implement it themselves using the existing unsafe Query methods, like above, and worry their use of unsafe will cause issues.

### PR DESCRIPTION
**Pull request by alice-i-cecile**
# Objective

- The inability to have multiple active mutable borrows into a query is a common source of borrow-checker pain for users.
- This is a pointless restriction if and only if we can guarantee that the entities they are accessing are unique.
- This could already by bypassed with `get_unchecked`, but that is an extremely unsafe API.
- Closes #2042.

## Solution

- Implement `Query::get_multiple` and `get_multiple_mut`, to provide access to the data of more than one entity at once.
- `get_multiple_mut` enforces uniqueness (to ensure soundness), but `get_multiple` does not (to reduce overhead).
- No equivalent methods were made on `World`, as multiple `EntityMut` cannot exist at once: the ability to add and remove components to entities with it make that design fundamentally unsound.
- Even when working in an exclusive world context, users can still create a `Query` and then use these methods.
- An iterator is returned, rather than a `Vec` or other collection in order to avoid allocations and ensure maximal flexibility.

## Breaking
Technically, this is a breaking change, as I've added a `AliasedMutability` variant to the `QueryEntityError` enum.

### CODE REVIEW

File: crates/bevy_ecs/src/system/query.rs [hunk 1]
```diff
@@ -640,6 +643,90 @@ where
...
+    /// # use bevy_ecs::prelude::*;
+    /// #[derive(Component, PartialEq)]
+    /// struct A(u64);
+    ///
+    /// let world = World::new();
+    /// let entity_1 = world.spawn().insert(A(1)).id();
+    /// let entity_2 = world.spawn().insert(A(2)).id();
+    ///
+    /// let a_query = world.query::<A>();
+    /// let a_iterator = name_query.get_multiple([entity_1, entity_2]);
```
Comment (MrGVSV) [inline crates/bevy_ecs/src/system/query.rs:666]:
```suggestion
    /// let a_iterator = a_query.get_multiple([entity_1, entity_2]);
```

File: crates/bevy_ecs/src/system/query.rs [hunk 2]
```diff
@@ -640,6 +643,90 @@ where
...
+    /// #[derive(Component, PartialEq)]
+    /// struct Name(String);
+    ///
+    /// let world = World::new();
+    /// let entity_1 = world.spawn().insert(A(1)).id();
+    /// let entity_2 = world.spawn().insert(A(2)).id();
+    /// let entity_3 = world.spawn().insert(A(3)).id();
+    ///
+    /// let a_query = world.query::<A>();
+    /// let a_iterator = name_query.get_multiple_mut(BTreeSet::from_iter([entity_1, entity_3]));
```
Comment (MrGVSV) [inline crates/bevy_ecs/src/system/query.rs:705]: Also here:

```suggestion
    /// let a_iterator = a_query.get_multiple_mut(BTreeSet::from_iter([entity_1, entity_3]));
```

File: crates/bevy_ecs/src/system/query.rs [hunk 3]
```diff
@@ -640,6 +682,208 @@ where
...
+
+        // Preallocating the HashSet used to check uniqueness based on the expected maximum amount of space
+        let (lower_bound, maybe_upper_bound) = entities_iter.size_hint();
+        let best_bound = if let Some(upper_bound) = maybe_upper_bound {
+            upper_bound
+        } else {
+            lower_bound
+        };
+
+        let entities_seen = HashSet::with_capacity(best_bound * std::mem::size_of::<Entity>());
```
Comment (Mathspy) [inline crates/bevy_ecs/src/system/query.rs:876]: If I am not misunderstanding the intent of this call (please correct me if I am, still new here!), I think we don’t need the size_of call as the method should be doing it internally at least that’s what I understood from [the documentation for with_capacity](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.with_capacity) [emphasis mine]
> The hash set will be able to hold at least `capacity` **elements** without reallocating.

```suggestion
        let entities_seen = HashSet::with_capacity(best_bound);
```

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/query.rs:876]: Ah, so you're right: it automatically looks at the size of its storage: I'd assumed that it was taking a size in bytes. Thanks!

File: crates/bevy_ecs/src/system/query.rs [hunk 4]
```diff
@@ -640,6 +682,208 @@ where
...
+        let entities_iter = entities.into_iter();
+
+        // Preallocating the HashSet used to check uniqueness based on the expected maximum amount of space
+        let (lower_bound, maybe_upper_bound) = entities_iter.size_hint();
+        let best_bound = if let Some(upper_bound) = maybe_upper_bound {
+            upper_bound
+        } else {
+            lower_bound
+        };
+
```
Comment (nakedible) [inline crates/bevy_ecs/src/system/query.rs:875]: Perhaps?
```suggestion
        let best_bound = maybe_upper_bound.unwrap_or(lower_bound);
```

File: crates/bevy_ecs/src/system/query.rs [hunk 5]
```diff
@@ -651,7 +895,7 @@ where
     /// this call does not result in multiple mutable references to the same component
     #[inline]
     pub unsafe fn get_unchecked(
-        &'s self,
+        &self,
```
Comment (nakedible) [inline crates/bevy_ecs/src/system/query.rs:898]: This is kind of unrelated fix in this pull, so maybe add one line in the "incidental" part of the pull description about it.

**Review by nakedible — CHANGES_REQUESTED** [review]
Community review! Looks good to me, but strained myself to get at least something to comment 😉

File: crates/bevy_ecs/src/system/query.rs (hunk 5 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/query.rs:893]: The elided lifetime should be the same as `'s`, so I don't think it's necessary to call out separately.

That said, I'd like to keep the explicit lifetime around for now; reasoning about this area of the code base's lifetimes was pretty gnarly.

**Comment by nakedible** [comment]
Also, probably just stating the obvious here, but adding an error to `QueryEntityError` would not be a breaking change if it would be marked as `#[non_exhaustive]`.

**Comment by alice-i-cecile** [comment]
We have accepted that every release will be extremely breaking, at least for now ;) I'm reasonably happy with causing compiler errors to draw attention to new behavior in this case.

**Comment by yilinwei** [comment]
For the lack of variadics, what about using and returning `const` generics? So `[Entity; N] -> [&mut T; N]`?

**Review by nakedible — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
> For the lack of variadics, what about using and returning const generics? So [Entity; N] -> [&mut T; N]?

IMO This is too restrictive, and will make this API more frustrating to use.

**Comment by alice-i-cecile** [comment]
Honestly I like the simpler more consistent API with const generics better. This is good, thanks @yilinwei!

**Review by yilinwei — APPROVED** [review]

File: crates/bevy_ecs/src/system/query.rs [hunk 6]
```diff
@@ -640,6 +678,99 @@ where
         }
     }
 
+    /// Returns the read-only query results for the iterator of [`Entity`]s provided.
+    ///
+    /// These values follow the order of your input iterator (if any).
```
Comment (MinerSebas) [inline crates/bevy_ecs/src/system/query.rs:683]:
```suggestion
    /// Returns the read-only query results for the provided Array of [`Entity`]s.
    ///
    /// These values follow the order of your input Array (if anyy).
```

File: crates/bevy_ecs/src/system/query.rs [hunk 7]
```diff
@@ -640,6 +678,99 @@ where
...
+    pub fn get_multiple<const N: usize>(
+        &'s self,
+        entities: [Entity; N],
+    ) -> impl Iterator<Item = Result<<Q::ReadOnlyFetch as Fetch>::Item, QueryEntityError>> {
+        entities.into_iter().map(|entity| self.get(entity))
+    }
+
+    /// Returns the query results for the [`BTreeSet`](std::collections::BTreeSet) of [`Entity`]s provided.
+    ///
+    /// These values follow the order of your input iterator (if any).
```
Comment (MinerSebas) [inline crates/bevy_ecs/src/system/query.rs:718]:
```suggestion
    /// Returns the query results for the provided Array of [`Entity`]s.
    ///
    /// These values follow the order of your input Array (if any).
```

File: crates/bevy_ecs/src/system/query.rs [hunk 8]
```diff
@@ -999,6 +1130,47 @@ where
...
+                    // SAFE: entities are checked for uniqueness using a HashSet
+                    unsafe { Query::<'w, 's, Q, F>::get_unchecked(self.query, entity) }
+                } else {
+                    Err(QueryEntityError::AliasedMutability)
+                }
+            })
+        } else {
+            self.index += 1;
+            None
+        }
```
Comment (MinerSebas) [inline crates/bevy_ecs/src/system/query.rs:1170]:
```suggestion
        } else {
            None
        }
```

File: crates/bevy_ecs/src/world/mod.rs [hunk 9]
```diff
@@ -20,6 +20,7 @@ use crate::{
 use std::{
     any::TypeId,
     fmt,
+    iter::IntoIterator,
```
Comment (MinerSebas) [inline crates/bevy_ecs/src/world/mod.rs:23]:
```suggestion
```
Seems like an forgotten change from the previous Version.

File: crates/bevy_ecs/src/system/query.rs [hunk 10]
```diff
@@ -275,6 +279,40 @@ where
...
+    /// let mut world = World::new();
+    /// world.spawn().insert(PowerLevel(9001));
+    ///
+    /// let query_state = world.query::<&PowerLevel>();
+    /// let query = Query::from_state(&mut world, &query_state);
+    /// let power_level = query.single();
+    ///
+    /// assert!(power_level.0 > 9000);
+    /// ```
+    pub fn from_state(world: &'w mut World, state: &'s QueryState<Q, F>) -> Self {
```
Comment (MinerSebas) [inline crates/bevy_ecs/src/system/query.rs:308]: This isn't used anywhere in this PR.

**Comment by alice-i-cecile** [comment]
@MinerSebas thanks for the review! I've fixed up your comments now.

**Comment by alice-i-cecile** [comment]
@MinerSebas, the `Query::from_state` method is actually needed to make these methods useful when we're working with a raw `World`. As a result, removing this made the doc-tests fail. IMO this should be part of the same PR; I'll clean it up.

File: crates/bevy_ecs/src/system/query.rs [hunk 11]
```diff
@@ -275,6 +279,41 @@ where
         }
     }
 
+    /// Creates a new [`Query`] from a [`QueryState`]
```
Comment (cart) [inline crates/bevy_ecs/src/system/query.rs:282]: It is important to call out that Query will not update QueryState to account for changes in World. If you create a QueryState, do something that creates a new archetype, then construct a Query from it, the new archetype will be ignored.

Using Query methods is essentially equivalent to the `QueryState::X_manual` methods.

File: crates/bevy_ecs/src/system/query.rs [hunk 12]
```diff
@@ -640,6 +679,99 @@ where
...
+    /// // Manually drop references so we can access the `World` again
+    /// std::mem::drop(a_iterator);
+    /// std::mem::drop(a_query);
+    ///
+    /// assert_eq!(*world.get::<A>(entity_1).unwrap(), A(11));
+    /// assert_eq!(*world.get::<A>(entity_2).unwrap(), A(2));
+    /// assert_eq!(*world.get::<A>(entity_3).unwrap(), A(33));
+    /// ```
+    #[inline]
+    pub fn get_multiple_mut<const N: usize>(
```
Comment (cart) [inline crates/bevy_ecs/src/system/query.rs:758]: Imo this shouldn't use an iterator or hash set. To avoid allocations and hashing, it should do a brute force in place entity collision check. For small numbers of entities (which the const generic basically enforces) this will be much much faster. I expect 99% of use cases to use 2 or 3 entities (one comparison and three comparisons, respectively). And it should return a fixed size array:

```rust
if let Ok([c1, c2, c3]) = query.get_multiple_mut([e1, e2, e3]) {
}
```

File: crates/bevy_ecs/src/system/query.rs [hunk 13]
```diff
@@ -640,6 +679,99 @@ where
...
+    /// #[derive(Component, PartialEq, Debug)]
+    /// struct A(u64);
+    ///
+    /// let mut world = World::new();
+    /// let entity_1 = world.spawn().insert(A(1)).id();
+    /// let entity_2 = world.spawn().insert(A(2)).id();
+    /// let entity_3 = world.spawn().insert(A(3)).id();
+    ///
+    /// let query_state = world.query::<&A>();
+    /// let a_query = Query::from_state(&mut world, &query_state);
```
Comment (cart) [inline crates/bevy_ecs/src/system/query.rs:703]: I don't think the QueryState -> Query pattern is something we should encourage in our docs.

Comment (cart) [reply crates/bevy_ecs/src/system/query.rs:703]: Query examples should just show systems using the query to avoid the need to show Query construction.

File: crates/bevy_ecs/src/system/query.rs [hunk 14]
```diff
@@ -640,6 +679,99 @@ where
...
+    /// let mut a_iterator = a_query.get_multiple([entity_3, entity_2, entity_1]).map(|i|i.unwrap());
+    /// assert_eq!(*a_iterator.next().unwrap(), A(3));
+    /// assert_eq!(*a_iterator.next().unwrap(), A(2));
+    /// assert_eq!(*a_iterator.next().unwrap(), A(1));
+    /// ```
+    #[inline]
+    pub fn get_multiple<const N: usize>(
```
Comment (cart) [inline crates/bevy_ecs/src/system/query.rs:710]: I think these should be implemented on QueryState with appropriate variants, then wired up to Query

File: crates/bevy_ecs/src/system/query.rs (hunk 12 above)
Comment (cart) [reply crates/bevy_ecs/src/system/query.rs:758]: Quick and dirty impl that ignores error handling:

```rust

    #[inline]
    pub fn get_multiple<const N: usize>(
        &'s self,
        entities: [Entity; N],
    ) -> Result<[<Q::ReadOnlyFetch as Fetch<'w, 's>>::Item; N], QueryEntityError> {
        for i in 0..entities.len() {
            for j in (i + 1)..entities.len() {
                if entities[i] == entities[j] {
                    panic!("cannot call on matching entities")
                }
            }
        }
        Ok(entities.map(|e| unsafe {
            self.state
                .get_unchecked_manual::<Q::ReadOnlyFetch>(
                    self.world,
                    e,
                    self.last_change_tick,
                    self.change_tick,
                )
                .unwrap()
        }))
    }
    ```

File: crates/bevy_ecs/src/system/query.rs (hunk 13 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/query.rs:703]: Sounds good to me.

File: crates/bevy_ecs/src/system/query.rs (hunk 14 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/query.rs:710]: Yep... I don't love this whole API duplication, but I think that's a reasonable call.

File: crates/bevy_ecs/src/system/query.rs (hunk 13 above)
Comment (cart) [reply crates/bevy_ecs/src/system/query.rs:703]: Doing that isn't too bad right now:

```rust
let mut system = IntoSystem::into_system(some_system);
system.initialize(world);
system.run((), world);
```

And we could / should hide those lines.

And we could add a `run_system(system, world)` function if we really wanted to (that initializes and runs the system now).

Comment (cart) [reply crates/bevy_ecs/src/system/query.rs:703]: (and calls IntoSystem)

File: crates/bevy_ecs/src/system/query.rs (hunk 14 above)
Comment (cart) [reply crates/bevy_ecs/src/system/query.rs:710]: I think it is necessary if we want "self driving" queries with direct world access (and I think we do).

File: crates/bevy_ecs/src/system/query.rs [hunk 15]
```diff
@@ -640,6 +679,99 @@ where
...
+        &'s self,
+        entities: [Entity; N],
+    ) -> impl Iterator<Item = Result<<Q::ReadOnlyFetch as Fetch>::Item, QueryEntityError>> {
```
Comment (BoxyUwU) [inline crates/bevy_ecs/src/system/query.rs:713]:
```suggestion
        &self,
        entities: [Entity; N],
    ) -> impl Iterator<Item = Result<<Q::ReadOnlyFetch as Fetch<'_, 's>>::Item, QueryEntityError>> {
```
Though to be honest I'm not sure there is a need for this API given the impl is so trivial :woman_shrugging:

File: crates/bevy_ecs/src/system/query.rs [hunk 16]
```diff
@@ -640,6 +679,99 @@ where
...
+    ///
+    /// assert_eq!(*world.get::<A>(entity_1).unwrap(), A(11));
+    /// assert_eq!(*world.get::<A>(entity_2).unwrap(), A(2));
+    /// assert_eq!(*world.get::<A>(entity_3).unwrap(), A(33));
+    /// ```
+    #[inline]
+    pub fn get_multiple_mut<const N: usize>(
+        &mut self,
+        entities: [Entity; N],
+    ) -> impl Iterator<Item = Result<<Q::Fetch as Fetch>::Item, QueryEntityError>> {
```
Comment (BoxyUwU) [inline crates/bevy_ecs/src/system/query.rs:761]:
```suggestion
    ) -> impl Iterator<Item = Result<<Q::Fetch as Fetch<'_, 's>>::Item, QueryEntityError>> {
```

File: crates/bevy_ecs/src/system/query.rs [hunk 17]
```diff
@@ -999,6 +1131,46 @@ where
...
+/// Iterator adaptor struct used for [`Query::get_multiple_mut`]('Query::get_multiple_mut`)
+/// See <https://stackoverflow.com/a/49813195> for more exposition
+struct GetMultipleMut<'w, 's, 'q, Q: WorldQuery, F: WorldQuery, const N: usize>
+where
+    F::Fetch: FilterFetch,
+{
+    // PERF: we could probably make this faster in common cases using a PetitSet
+    // or other set optimized for small sizes
+    seen: HashSet<Entity>,
+    query: &'q Query<'w, 's, Q, F>,
```
Comment (BoxyUwU) [inline crates/bevy_ecs/src/system/query.rs:1143]:
```suggestion
    query: &'q mut Query<'w, 's, Q, F>,
```

File: crates/bevy_ecs/src/system/query.rs [hunk 18]
```diff
@@ -999,6 +1131,46 @@ where
...
+    index: usize,
+    array: [Entity; N],
+}
+
+impl<'w, 's, 'q, Q: WorldQuery, F: WorldQuery, const N: usize> Iterator
+    for GetMultipleMut<'w, 's, 'q, Q, F, N>
+where
+    F::Fetch: FilterFetch,
+{
+    type Item = Result<<Q::Fetch as Fetch<'w, 's>>::Item, QueryEntityError>;
```
Comment (BoxyUwU) [inline crates/bevy_ecs/src/system/query.rs:1153]:
```suggestion
    type Item = Result<<Q::Fetch as Fetch<'q, 's>>::Item, QueryEntityError>;
```

**Review by BoxyUwU — CHANGES_REQUESTED** [review]
You should add a compile fail test making sure that you cant make a `GetMultipleMut` iterator and then call methods on `Query` and then use the results from `GetMultipleMut` to get an aliased mutable ref

**Comment by BoxyUwU** [comment]
Lol those errors are not super helpful... can be fixed by replacing `impl Iterator` with the actual iterator type and by doing `Query::<'q, 's, Q, F>::get_unchecked`

**Comment by alice-i-cecile** [comment]
Closing in favor of the less messy #4298.
